File: workspace/data_mart.py
import os
import sys
import argparse
import logging
from config import GreenplumConfig, spark_conf_manager
from spark_manager import SparkManager, SparkProfile
from pyspark.ml.feature import VectorAssembler, StandardScaler
from pyspark.ml.functions import vector_to_array

# Настройка логирования для DataMart
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger("DataMart")

class DataMart:
    def __init__(self):
        logger.info("Initializing SparkSession for DataMart")
        self.spark = SparkManager().get_session(SparkProfile.DATA_MART)
        
        self.jdbc_url = GreenplumConfig.JDBC_URL
        self.conn_properties = GreenplumConfig.PROPERTIES

        # Берем названия таблиц из единого конфигуратора
        self.raw_table = GreenplumConfig.RAW_TABLE
        self.features_table = GreenplumConfig.FEATURES_TABLE
        self.predictions_table = GreenplumConfig.PREDICTIONS_TABLE
        self.clustered_table = GreenplumConfig.CLUSTERED_TABLE

    def extract_and_transform(self):
        logger.info("=== [DataMart] Запуск извлечения и предобработки данных ===")
        logger.info(f">>> Подключение к Greenplum по адресу: {self.jdbc_url}")

        try:
            raw_df = self.spark.read.jdbc(self.jdbc_url, self.raw_table, properties=self.conn_properties)
        except Exception as e:
            logger.error(f"Ошибка чтения из БД: {e}")
            sys.exit(1)

        logger.info(">>> Очистка данных от пропусков (NaN)...")
        clean_df = raw_df.dropna()

        logger.info(">>> Векторизация признаков...")
        assembler = VectorAssembler(
            inputCols=["energy-kcal_100g", "proteins_100g", "fat_100g"],
            outputCol="rawFeatures",
            handleInvalid="skip"
        )
        vectorized_df = assembler.transform(clean_df)

        logger.info(">>> Масштабирование признаков...")
        scaler = StandardScaler(
            inputCol="rawFeatures",
            outputCol="features",
            withStd=True,
            withMean=True
        )
        scaled_df = scaler.fit(vectorized_df).transform(vectorized_df)

        # Преобразуем VectorUDT в обычный массив (ArrayType) для сохранения в Greenplum
        final_features_df = scaled_df.select("id", vector_to_array("features").alias("features"))

        # Load to Database
        logger.info(f">>> Сохранение предобработанных данных в таблицу {self.features_table}...")
        final_features_df.write \
            .mode("overwrite") \
            .option("createTableOptions", "DISTRIBUTED BY (id)") \
            .jdbc(self.jdbc_url, self.features_table, properties=self.conn_properties)
            
        logger.info("=== [DataMart] Данные успешно подготовлены и выгружены в БД! ===")

    def load_results(self):
        logger.info("=== [DataMart] Запуск загрузки результатов в Greenplum ===")
        logger.info(f">>> Чтение результатов кластеризации из таблицы {self.predictions_table}...")
        
        try:
            results_df = self.spark.read.jdbc(self.jdbc_url, self.predictions_table, properties=self.conn_properties)
        except Exception as e:
            logger.error(f"Ошибка чтения результатов модели из БД: {e}")
            sys.exit(1)
        
        logger.info(">>> Предпросмотр полученных данных от модели:")
        results_df.show(5)

        logger.info(f">>> Запись итоговых данных в таблицу {self.clustered_table}...")
        
        # Пишем в Greenplum с учетом DISTRIBUTED BY (id)
        results_df.write \
            .mode("overwrite") \
            .option("createTableOptions", "DISTRIBUTED BY (id)") \
            .jdbc(self.jdbc_url, self.clustered_table, properties=self.conn_properties)

        logger.info("=== [DataMart] Данные успешно загружены! Контур завершил работу. ===")

    def stop(self):
        self.spark.stop()

def main():
    parser = argparse.ArgumentParser(description="DataMart ETL Controller")
    parser.add_argument("--mode", choices=["extract", "load"], required=True, 
                        help="Режим работы: 'extract' для выгрузки из БД, 'load' для записи результатов")
    args = parser.parse_args()

    datamart = DataMart()
    if args.mode == "extract":
        datamart.extract_and_transform()
    elif args.mode == "load":
        datamart.load_results()
        
    datamart.stop()

if __name__ == "__main__":
    main()

File: workspace/datamart/src/main/scala/DataMart.scala
import org.apache.spark.sql.{SparkSession, SaveMode}
import org.apache.spark.ml.feature.{VectorAssembler, StandardScaler}
import org.apache.spark.ml.functions.vector_to_array
import org.apache.spark.sql.functions.col

import java.util.Properties
import java.io.File
import com.typesafe.config.ConfigFactory
import scala.collection.JavaConverters._

object DataMart {
  def main(args: Array[String]): Unit = {
    System.setProperty("java.net.preferIPv4Stack", "true")

    if (args.length < 2 || args(0) != "--mode") {
      println("[ERROR] Использование: DataMart --mode <extract|load>")
      sys.exit(1)
    }
    val mode = args(1)

    println(">>> Чтение конфигурации из spark.conf...")
    val rootConfig = ConfigFactory.parseFile(new File("spark.conf")).resolve()
    val config = rootConfig.getConfig("datamart")

    // Извлекаем настройки БД
    val dbConfig = config.getConfig("db")
    val jdbcUrl = s"jdbc:postgresql://${dbConfig.getString("host")}:${dbConfig.getString("port")}/${dbConfig.getString("db_name")}?sslmode=disable"
    
    val connectionProperties = new Properties()
    connectionProperties.put("user", dbConfig.getString("user"))
    connectionProperties.put("password", dbConfig.getString("password"))
    connectionProperties.put("driver", "org.postgresql.Driver")

    val tables = config.getConfig("tables")
    val rawTable = tables.getString("raw")
    val featuresTable = tables.getString("features")
    val predictionsTable = tables.getString("predictions")
    val clusteredTable = tables.getString("clustered")

    // 2. Инициализация сессии Spark
    val sparkConfig = config.getConfig("spark")
    val builder = SparkSession.builder()
      .appName(s"${sparkConfig.getString("app_name")}_${mode.capitalize}")
      .master(sparkConfig.getString("master"))
      .config("spark.driver.bindAddress", "0.0.0.0")
      .config("spark.driver.host", "127.0.0.1")

    if (sparkConfig.hasPath("settings")) {
      val settings = sparkConfig.getConfig("settings")
      settings.entrySet().asScala.foreach { entry =>
        // Ключи в HOCON парсятся в кавычках, unwrapped() снимает их
        builder.config(entry.getKey, entry.getValue.unwrapped().toString)
      }
    }

    val spark = builder.getOrCreate()
    spark.sparkContext.setLogLevel("WARN")

    mode match {
      case "extract" => 
        extractAndTransform(spark, jdbcUrl, connectionProperties, rawTable, featuresTable)
      case "load" => 
        loadResults(spark, jdbcUrl, connectionProperties, predictionsTable, clusteredTable)
      case _ =>
        println(s"[ERROR] Неизвестный режим работы: $mode. Используйте 'extract' или 'load'.")
        sys.exit(1)
    }

    spark.stop()
  }

  // РЕЖИМ 1: Извлечение и Предобработка (EXTRACT)
  def extractAndTransform(spark: SparkSession, url: String, props: Properties, rawTable: String, featuresTable: String): Unit = {
    println("=== [DataMart] Запуск извлечения и предобработки данных (Scala) ===")
    println(s">>> Чтение сырых данных из Greenplum (таблица $rawTable)...")
    
    val rawDf = spark.read.jdbc(url, rawTable, props)
    val cleanDf = rawDf.na.drop()

    println(">>> Векторизация и масштабирование признаков...")
    val assembler = new VectorAssembler()
      .setInputCols(Array("energy-kcal_100g", "proteins_100g", "fat_100g"))
      .setOutputCol("rawFeatures")
      .setHandleInvalid("skip")
    
    val vectorizedDf = assembler.transform(cleanDf)

    val scaler = new StandardScaler()
      .setInputCol("rawFeatures")
      .setOutputCol("features")
      .setWithStd(true)
      .setWithMean(true)
      
    val scalerModel = scaler.fit(vectorizedDf)
    val scaledDf = scalerModel.transform(vectorizedDf)

    val finalFeaturesDf = scaledDf.select(col("id"), vector_to_array(col("features")).alias("features"))

    println(s">>> Сохранение фичей для ML-модели в таблицу $featuresTable (Greenplum)...")
    finalFeaturesDf.write
      .mode(SaveMode.Overwrite)
      .option("createTableOptions", "DISTRIBUTED BY (id)")
      .jdbc(url, featuresTable, props)

    println("=== [DataMart] Данные успешно подготовлены и выгружены в БД! ===")
  }

  // РЕЖИМ 2: Загрузка результатов кластеризации (LOAD)
  def loadResults(spark: SparkSession, url: String, props: Properties, predictionsTable: String, clusteredTable: String): Unit = {
    println("=== [DataMart] Запуск загрузки результатов в Greenplum (Scala) ===")
    println(s">>> Чтение предсказаний модели из промежуточной таблицы $predictionsTable...")
    
    val resultsDf = spark.read.jdbc(url, predictionsTable, props)
    
    println(">>> Предпросмотр полученных данных от модели:")
    resultsDf.show(5)

    println(s">>> Запись финальных результатов в итоговую таблицу $clusteredTable...")
    resultsDf.write
      .mode(SaveMode.Overwrite)
      .option("createTableOptions", "DISTRIBUTED BY (id)")
      .jdbc(url, clusteredTable, props)

    println("=== [DataMart] Данные успешно загружены! Контур завершил работу. ===")
  }
}